Add "info" option to display detail information of a sport session.

ImfeldC opened the pull request:
Here my first "pull request", to test general workflow.
Is this correct that way?

ImfeldC committed:
Add "info" option to display detail information of a sport session.

fcrespel commented on src/main/java/me/crespel/runtastic/RuntasticExportConverter.java:
```diff
@@ -61,7 +69,25 @@ protected void doList(File path) throws FileNotFoundException, IOException {
+
+	protected void doInfo(File path, String id) throws FileNotFoundException, IOException {
+		SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+		List<SportSession> sessions = converter.listSportSessions(path);
+		for (SportSession session : sessions) {
+			if ( session.getId().equalsIgnoreCase(id) ) {
```
Since we're only interested in 1 session here, it would be more efficient to read and parse only that one, instead of the whole export folder. I would suggest adding a `getSportSession(File path, String id)` method to the `ExportConverter` class.

ImfeldC replied:
I'll fix it for next commit.

fcrespel commented on src/main/java/me/crespel/runtastic/RuntasticExportConverter.java:
```diff
@@ -61,7 +69,25 @@ protected void doList(File path) throws FileNotFoundException, IOException {
+				System.out.println(sdf.format(session.getStartTime()) + " - ID: " + session.getId() );
+				System.out.println("      Sport Type: " + session.getSportTypeId() + ", Surface Type: " + session.getSurfaceId() + ", Feeling Id: " + session.getSubjectiveFeelingId());
+				System.out.println("      Duration: " + session.getDuration() + " ms / " + session.getDuration()/1000 + " s / " + session.getDuration()/60000 + " min");
+				System.out.println("      Distance: " + session.getDistance().floatValue()/1000 + " km, Calories: " + session.getCalories());
+				System.out.println("      Avg Pace: " + session.getDurationPerKm().floatValue()/60000 + " min/km");
```
Instead of calling `.floatValue()` on these Integers, you could just divide them by 1000.0 (i.e. a double) or 60000.0.

But careful! some sessions may have no distance (no GPS) or duration per km, so these two lines can throw a NullPointerException at runtime. This can be fixed by checking if the value is not null before displaying the line.

ImfeldC replied:
Will be fixed in next commit.

fcrespel commented on src/main/java/me/crespel/runtastic/RuntasticExportConverter.java:
```diff
@@ -61,7 +69,25 @@ protected void doList(File path) throws FileNotFoundException, IOException {
+			if ( session.getId().equalsIgnoreCase(id) ) {
+				System.out.println(sdf.format(session.getStartTime()) + " - ID: " + session.getId() );
+				System.out.println("      Sport Type: " + session.getSportTypeId() + ", Surface Type: " + session.getSurfaceId() + ", Feeling Id: " + session.getSubjectiveFeelingId());
+				System.out.println("      Duration: " + session.getDuration() + " ms / " + session.getDuration()/1000 + " s / " + session.getDuration()/60000 + " min");
```
Possible improvement: display the duration as hours:minutes:seconds for easier reading.

ImfeldC replied:
Removed the [ms] and [s] values; I think showing [min] is good enough.

ImfeldC replied:
Will be fixed in next commit.

fcrespel commented on src/main/java/me/crespel/runtastic/RuntasticExportConverter.java:
```diff
@@ -61,7 +69,25 @@ protected void doList(File path) throws FileNotFoundException, IOException {
 		List<SportSession> sessions = converter.listSportSessions(path);
 		for (SportSession session : sessions) {
-			System.out.println(sdf.format(session.getStartTime()) + " - ID: " + session.getId() + ", type: " + session.getSportTypeId() + ", duration: " + session.getDuration() + "s");
+			System.out.println(sdf.format(session.getStartTime()) + " - ID: " + session.getId() + ", Sport Type: " + session.getSportTypeId() + ", duration: " + session.getDuration()/60000 + "min");
```
Possible improvement: display the duration as hours:minutes:seconds for easier reading.

ImfeldC replied:
Would be useful and nice. I don't know an "easy" fix.

ImfeldC replied:
Found solution, use new Time(session.getDuration()).toString()

fcrespel commented:
Thanks for the contribution! This is the way ;-)
I've added some comments in the code review, please let me know if you'd like to handle them. Any further commit you do will automatically be added to this PR.

ImfeldC committed:
Incorporate review feedback.

ImfeldC commented:
Changes published back.

ImfeldC committed:
Change output of duration (use new Time(session.getDuration()).toString())

ImfeldC commented:
All review findings incorporated.

fcrespel commented on src/main/java/me/crespel/runtastic/converter/ExportConverter.java:
```diff
@@ -38,6 +38,18 @@
+	public SportSession getSportSession(File path, String id) throws FileNotFoundException, IOException {
+		path = normalizeSportSessionPath(path);
+		List<SportSession> sessions = new ArrayList<>();
+		File[] files = path.listFiles(file -> file.getName().endsWith(id +".json"));
```
More simply, since you already have the path and ID, you don't need to list files: you can just do a `new File(path, id + ".json"), parse it and return the result.

ImfeldC replied:
Good point! Load file now directly: new File(path,id +".json")

fcrespel commented on src/main/java/me/crespel/runtastic/RuntasticExportConverter.java:
```diff
@@ -61,7 +70,23 @@ protected void doList(File path) throws FileNotFoundException, IOException {
 		List<SportSession> sessions = converter.listSportSessions(path);
 		for (SportSession session : sessions) {
-			System.out.println(sdf.format(session.getStartTime()) + " - ID: " + session.getId() + ", type: " + session.getSportTypeId() + ", duration: " + session.getDuration() + "s");
+			System.out.println(sdf.format(session.getStartTime()) + " - ID: " + session.getId() + ", Sport Type: " + session.getSportTypeId() + ", duration: " + new Time(session.getDuration()).toString());
```
I get a weird result with `new Time(session.getDuration()).toString())` : my activities are displayed with +1 hour (e.g. 12 mins becomes 1:12:00). Working with dates and durations in Java is a bit of a pain, I'd suggest finding alternate solutions on StackOverflow.

ImfeldC replied:
Fixed, use now: Duration.ofMillis(session.getDuration()).toString()

fcrespel reviewed (changes requested):
Getting better! there are still two points to address, see comments (also, sorry for the delay).

ImfeldC committed:
Resolve, review findings ..
- Load file now directly: new File(path,id +".json")
- use now: Duration.ofMillis(session.getDuration()).toString()

fcrespel reviewed (approved):
OK, I'll proceed with the merge :-)

fcrespel commented:
Note that I further simplified `ExportConverter.getSportSession()` in an additional commit after the merge. Thanks for your work! 👍

ImfeldC commented:
Cool! Thanks for accepting PR ... my official "first" contribution on github :-)
I update my fork with your additional commit. Be prepared for the next PR; this contains the real changes (and enhancments) ....
